File: day1/ruby/part_2.rb
input_file = "input.txt"

left_list = []
right_list = []

File.foreach(input_file) do |line|
  left, right = line.split.map(&:to_i)
  left_list << left
  right_list << right
end

right_counts = right_list.tally

similarity_score = left_list.sum { |num| num * (right_counts[num] || 0) }

puts similarity_score

File: day1/typescript/part_2.ts
import * as fs from 'fs';

const inputFile = "input.txt";

const fileContent = fs.readFileSync(inputFile, 'utf-8');
const lines = fileContent.trim().split('\n');

const leftList: number[] = [];
const rightList: number[] = [];

lines.forEach(line => {
    const [left, right] = line.split(' ').map(Number);
    leftList.push(left);
    rightList.push(right);
});

const rightCounts: Record<number, number> = {};
rightList.forEach(num => {
    rightCounts[num] = (rightCounts[num] || 0) + 1;
});

let similarityScore = 0;
leftList.forEach(num => {
    similarityScore += num * (rightCounts[num] || 0);
});

console.log(similarityScore);
